interfaces/cython/cantera/composite.py

```diff
@@ -528,32 +528,46 @@ def __init__(self, phase, shape=(0,), states=None, extra=None, meta=None):
         reserved = self.__dir__()
 
         self._extra = OrderedDict()
+
         if isinstance(extra, dict):
             for name, v in extra.items():
                 if name in reserved:
                     raise ValueError(
                         "Unable to create extra column '{}': name is already "
                         "used by SolutionArray objects.".format(name))
                 if not np.shape(v):
-                    self._extra[name] = [v]*self._shape[0]
-                elif len(v) == self._shape[0]:
-                    self._extra[name] = list(v)
+                    self._extra[name] = np.full(self._shape, v)
+                elif np.array(v).shape == self._shape:
+                    self._extra[name] = np.array(v)
                 else:
-                    raise ValueError("Unable to map extra SolutionArray"
-                                     "input for named {!r}".format(name))
+                    raise ValueError("Unable to map extra SolutionArray "
+                                     "input named {!r}".format(name))
+        elif extra is not None:
+            if isinstance(extra, np.ndarray):
+                extra = extra.flatten()
+            elif isinstance(extra, str):
```

Comment:
I just noticed this possibility being considered: please be aware that this may break things elsewhere (e.g. restoration of CSV data to `SolutionArray`). I.e. while I haven’t run any tests, specifically planning for `str` input may require additional ‘fixes’)

Reply:
Other than possibly breaking CSV/HDF input/output, is there a reason to prohibit strings? Would it be better to raise that error in the CSV/HDF input/output methods, if there are no other problems with using strings as extra items?

Reply:
Actually, this is just checking the name of the extra items, not the values. So this code won't need to change at all.

Reply:
As mentioned, I had never considered str to be a sanctioned use in the first place.

Reply:
For the names of the columns? In that case, I'd say that the HDF/CSV code needs to change, if it is broken.

Reply:
No, values - at this point, this can be considered resolved here, as I opened an issue in #896